### ISSUE TITLE
Optionally allow function updates

### ISSUE THREAD

**Issue by cybrown** [AFASSoftware/maquette#23]
Before all, I'm not aware of all performance issues, so take the associated PR as an example of what can be done, not the way to do it. Also, it needs an option in the maquette.createProjection factory to be effective.

So I've made it possible to change the event handlers. It seems that this modification does not have a lot of consequences on performances, it adds a property on the DOM Node to store the original handler, so eventHandlerInterceptor is called only once.

In case of the PR associated with this issue is accepted, I'll sign the CLA.

**Comment by peterwestendorp** [comment]
Hey @cybrown, thanks for taking the time and effort to improve Maquette! I think I understand what you're trying to do, but could you elaborate a bit on what you're trying to achieve? What problem are you trying to solve? (A little Gist or Fiddle would be helpful). Just to make sure we're both talking about the same thing...

**Comment by cybrown** [comment]
Hi,

I've made two examples on jsbin, I hope they'll describe what I want to achive !

* First using the original maquette.js (I had to find a cdn, found out that the file in the git repo is in ts, nice :) !)
https://jsbin.com/nuzolenedu/edit?html,output

* Now with my fork
https://jsbin.com/donucakoyu/edit?html,output

**Comment by RickHoving** [comment]
Hi,

I've looked at your samples, please correct me if I misunderstood you.
We've deliberately disabled updating functions on subsequent renders.
The reason for this is because on every render a new instance of the anonymous function is created.
Because the render function is triggered so many times, this would be a performance issue if you'd create a bunch of anonymous functions.
Also take into account the additional work of the garbage collector and such.
Therefore we disabled the functionality.

**Comment by cybrown** [comment]
Yes you are correct. That's why I've made it disabled by default.
I understand that this can be a performance issue, but I am not aware of this limitation in other vdom librairies.

**Comment by johan-gorter** [comment]
Hello @cybrown,

I am aware that other virtual DOM libraries allow you to use anonymous functions. Some libraries have made their own event-abstraction layer and others just update the DOM on every render.

Maquette was designed to be simple, lightweight and performant. An event-abstraction layer would make maquette harder to learn and make it less lightweight. Updating the DOM on every render would sacrifice performance, although I never benchmarked the cost of this.

In our experience, using inline anonymous functions in the DOM clutters up the hyperscript, especially when applications and components are growing more complex. (I am aware that using ES2015 or Typescript reduce this clutter a little bit.) We think it is a best practice to extract the event handlers and we would encourage others to do this as well. This makes the hyperscript easier to read and it has performance benefits, because it does not needlessly recreate Function objects.

Making it optional is tempting, because that would attract a larger audience and allow innovation. There is one downside however. We are trying hard to keep maquette lightweight (3Kb) and adding options like this adds more code. This is why I would like to postpone adding this option until more people are asking for it. Meanwhile, feel free to use your fork to gain more experience with this option, and maybe do some benchmarking with it.

By the way: we are currently in the process of migrating from javascript to typescript, hence the merge conflicts. Sorry about this. The last released version was still using javascript.

**Comment by cybrown** [comment]
Thank you very much for this explanation. This is exactly why this project is great ! I wish the transition to typescript will be OK.

**Comment by katywings** [comment]
I am interested in this option!

Two use cases:

1. The static eventhandlers dont have access to the this or some other relevant context
    The problem in this scenario is, that you have to apply the relevant context variables to element properties. Now this is just something which feels strange to me - it reminds me of template languages ;).

2. I never thought about this, but lately we came into this situation: we have something like a plugin system. Each plugin has a render function which gets called from the plugin system. This render functio  has a helpers parameter and one of the helpers is a tinymce helper: basically a maquette function which creates a textarea and inits tinymce on it. Our problem was that we needed to customize the tinymces onchange eventhandler.
The eventhandler is like:
- getValue
- getState
- apply Value to default location in state
    Or if customizer exists: state = customizer (state, value)
- dispatch

We figured out a solution without anonymous eventhandlers but that solution is frustrating: The plugin needs to have an own element afterCreate eventhandler which then calls the original helpers afterCreate with the customizer as parameter. We needed to rewrite the whole thing multible times because with every earlier iteration we somewhere stumbled upon an anonymous function.

I think it would be great for maquette to have the anonymous functions optional with the con of slower performance!

**Comment by katywings** [comment]
Hey @johan-gorter 

Wow! I didn't know about `.bind`. Hell yeah, thats exactly what I would have needed :+1: 

This codepen describes what I was trying to do: https://codepen.io/katywings/pen/zwWrvq

I think this should be mentioned in the tutorial / three rules pages! Everywhere is mentioned that anonymous event handlers can't be used for performance reasons, but that there exists a common tool in form of bind as property to solve most of the use cases, isnt mentioned - thats not ideal self-promotion :laughing:  

Anyway: thank you a lot for the great tip!
